pages/dashboard.js

```diff
@@ -63,10 +63,102 @@ import UniversalBenefitCard from '../components/molecules/UniversalBenefitCard'
-import GetEIProgramData from './api/programData/ei'
+import { useCallback, useEffect, useState } from 'react'
 
 export default function Dashboard(props) {
+  const [advertisingCards, setAdvertisingCards] = useState(
+    props.advertisingCards
+  )
+  const [usersBenefits, setUsersBenefits] = useState([])
+  const [cppLoading, setCppLoading] = useState(false)
+  const [eiLoading, setEiLoading] = useState(false)
+  //no requirements to do anything with the errors yet
+  const [cppError, setCppError] = useState(false)
+  const [eiError, setEiError] = useState(false)
+
+  const FilterAdvertisingCards = useCallback(
+    (benefits) => {
+      //determin which advertising cards to remove
+      const cardsToRemove = []
+      if (benefits && benefits.length) {
+        if (
+          !CheckAllBenefitsForAdvertising(
+            benefits,
+            [StatusCodes.inPayment, StatusCodes.Pending],
+            ProgramCodes.CPP,
+            TypeCodes.CPPRetirement
+          )
+        ) {
+          cardsToRemove.push(APPLICATION_CARD_CPP)
+        }
+        if (
+          !CheckAllBenefitsForAdvertising(
+            benefits,
+            [StatusCodes.inPayment, StatusCodes.Pending],
+            ProgramCodes.OAS
+          )
+        ) {
+          cardsToRemove.push(APPLICATION_CARD_OAS)
+        }
+        if (
+          !CheckAllBenefitsForAdvertising(
+            benefits,
+            [StatusCodes.inPayment, StatusCodes.Pending],
+            ProgramCodes.GIS
+          )
+        ) {
+          cardsToRemove.push(APPLICATION_CARD_GIS)
+        }
+        if (
+          !CheckAllBenefitsForAdvertising(
+            benefits,
+            [StatusCodes.inPayment, StatusCodes.Pending],
+            ProgramCodes.CPP,
+            TypeCodes.CPPDisability
+          )
+        ) {
+          cardsToRemove.push(APPLICATION_CARD_CPPD)
+        }
+      }
+
+      //now remove them if they should be removed
+      return cardsToRemove.length
+        ? props.advertisingCards.filter((c) =>
+            cardsToRemove.find((r) => r.benefitType !== c.benefitType)
+          )
+        : props.advertisingCards
+    },
+    [props.advertisingCards]
+  )
+
+  const fetchProgramData = useCallback(
+    (program, setLoading, setError) => {
+      setLoading(true)
+      fetch(`api/programData/${program}`)
+        .then((res) => res.json())
+        .then((data) => {
+          const currData = usersBenefits
+          currData.push(data)
+          setUsersBenefits(currData)
+          setAdvertisingCards(FilterAdvertisingCards(currData))
+          setLoading(false)
+        })
+        .catch((error) => {
+          setLoading(false)
+          setError(error)
+        })
+    },
+    [usersBenefits, FilterAdvertisingCards]
+  )
+
+  useEffect(() => {
+    fetchProgramData('cpp', setCppLoading, setCppError)
+  }, [fetchProgramData])
+
+  useEffect(() => {
+    fetchProgramData('ei', setEiLoading, setEiError)
+  }, [fetchProgramData])
```

Comment:
The useEffect can be combined into one with empty brackets at the end ie [fetchProgramData] -> []
The app should only make this request once and I'm concerned about the behaviour of setting it to a call back.

Ideally useEffects should be triggered once on page load or multiple times triggered by updated state

Reply:
It could be change it to call both API's in one `useEffect`, but thought you had suggested it be broken into two?

The brackets with `[fetchProgramData]` is required to identify the dependency of that function. And that function is using a `useCallback' because it's also required for compile as the function is referenced in the useEffect. Changing the useCallback function to be a regular function does cause the recursive infinite loop of API requests (or adding duplicate from the result - I didn't debug as to which it was)

Reply:
Take some time to read over the ReactJS documentation. Specifically the React hook documentation on useState and useEffect.

There are also examples of querying using useState and useEffect in scrum poker

Reply:
So reading the docs, and looking at this change, I'm still not fully clear on what your suggesting. That being said, I've pushed a change that moves both API requests into one useEffect and the API fetch into an async function within it. (The issue I think you are trying to point out being that they are both updating a dependency of `userBenefits` of each other, so they could keep re-triggering the other once one completes).

I did leave the FilterAdvertisingCards as a useCallback, this seems to be agreeable to the docs; and everything else I tried broke the app.

Reply:
Merging as code is needed for other sprint jira tickets